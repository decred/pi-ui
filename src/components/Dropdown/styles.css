.dropdownWrapper {
  position: relative;
  align-self: flex-start;
  display: inline-flex;
  flex-direction: column;
}

.headerWrapper {
  display: flex;
  align-items: center;
}

.headerWrapper:hover {
  cursor: pointer;
}

.dropdownHeader {
  display: flex;
  margin-right: 0.6rem;
  color: var(--color-primary-dark);
}

.arrowAnchor {
  position: relative;
  width: 20px;
}

.customArrowAnchor {
  position: absolute;
}

.arrowAnchor:after {
  content: "";
  width: 0;
  height: 0;
  border: 7px solid transparent;
  border-color: var(--color-gray) transparent transparent transparent;
  position: absolute;
  top: -4px;
  right: 2px;
}

.arrowAnchor:hover:after {
  border-color: var(--color-primary) transparent transparent transparent;
}

.arrowAnchor.open:after {
  top: -11px;
  border-color: transparent transparent var(--color-primary) transparent;
}

.dropdownList {
  z-index: 100;
  position: absolute;
  background-color: #fff;
  display: inline-flex;
  flex-direction: column;
  justify-content: space-around;
  white-space: pre;
  align-items: flex-end;
  margin-top: 3.5rem;
  padding: 0 1.5rem 2rem 1.5rem !important;
  -webkit-box-shadow: 0px 10px 20px 0px rgba(0, 0, 0, 0.16);
  -moz-box-shadow: 0px 10px 20px 0px rgba(0, 0, 0, 0.16);
  box-shadow: 0px 10px 20px 0px rgba(0, 0, 0, 0.16);
}

.dropdownItem {
  list-style-type: none;
  display: flex;
  overflow: auto;
  text-align: left;
  font-size: var(--font-size-normal);
  color: var(--tab-font-color);
  padding: 1.75rem 0 1.75rem 0 !important;
}

.dropdownItem:hover {
  cursor: pointer;
  color: var(--tab-hover-color) !important;
}
