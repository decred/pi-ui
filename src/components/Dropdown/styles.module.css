.dropdownWrapper {
  position: relative;
  align-self: flex-start;
  display: inline-flex;
}

.headerWrapper {
  display: flex;
  align-items: center;
}

.headerWrapper:hover {
  cursor: pointer;
}

.dropdownHeader {
  display: flex;
  margin-right: 0.8rem;
  color: var(--color-primary-dark);
}

.arrowAnchor {
  position: relative;
  cursor: pointer;
  margin-left: 1.25rem;
}

.arrowAnchor:after {
  content: "";
  width: 0;
  height: 0;
  border: 7px solid transparent;
  border-color: var(--dropdown-arrow-color) transparent transparent transparent;
  position: absolute;
  top: -4px;
  right: 2px;
}

.arrowAnchor:hover:after {
  border-color: var(--color-primary) transparent transparent transparent;
}

.arrowAnchor.open:after {
  top: -11px;
  border-color: transparent transparent var(--color-primary) transparent;
}

.headerWrapper:hover .arrowAnchor:not(.open):after {
  border-color: var(--color-primary) transparent transparent transparent;
}

.dropdownList {
  z-index: 100;
  position: absolute;
  background-color: var(--card-background);
  display: inline-flex;
  left: 50%;
  transform: translateX(-50%);
  flex-direction: column;
  justify-content: space-around;
  white-space: pre;
  align-items: flex-end;
  margin-top: 3.5rem;
  padding: 0 1.5rem 2rem 1.5rem !important;
  -webkit-box-shadow: 0px 10px 20px 0px rgba(0, 0, 0, 0.16);
  -moz-box-shadow: 0px 10px 20px 0px rgba(0, 0, 0, 0.16);
  box-shadow: 0px 10px 20px 0px rgba(0, 0, 0, 0.16);
}

.dropdownItem {
  list-style-type: none;
  display: flex;
  overflow: auto;
  text-align: left;
  font-size: var(--font-size-normal);
  color: var(--tab-text-color);
  padding: 1.75rem 0 1.75rem 0 !important;
}

.dropdownItem:hover {
  cursor: pointer;
  color: var(--dropdown-item-hover-color) !important;
}
