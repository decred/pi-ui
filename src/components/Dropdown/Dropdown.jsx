import React, { useState, useCallback } from "react";
import PropTypes from "prop-types";
import { animated, useTransition } from "react-spring";
import useClickOutside from "../../hooks/useClickOutside";
import { classNames } from "../../utils";
import styles from "./styles.css";

const DefaultTrigger = ({ onClick, title, open }) => (
  <div className={styles.headerWrapper} onClick={onClick}>
    <span className={styles.dropdownHeader}>{title}</span>
    <div className={classNames(styles.arrowAnchor, open && styles.open)} />
  </div>
);

const Dropdown = ({
  children,
  className,
  title,
  itemsListClassName,
  onDropdownClick,
  show,
  closeOnOutsideClick,
  closeOnItemClick,
  customDropdownTrigger,
  dropdownArrowClassName,
  ...props
}) => {
  const [innerStateShow, setInnerStateShow] = useState(false);
  const dropdownOpenned = show || innerStateShow;
  const closeDropdown = useCallback(() => {
    setInnerStateShow(false);
  }, [setInnerStateShow]);

  const closeDropdownHandler =
    closeOnOutsideClick && dropdownOpenned
      ? onDropdownClick || closeDropdown
      : () => null;
  const [dropdownRef] = useClickOutside(closeDropdownHandler);

  const handleTriggerClick = useCallback(() => {
    if (!onDropdownClick) {
      setInnerStateShow(!innerStateShow);
      return;
    }
    onDropdownClick();
  }, [setInnerStateShow, onDropdownClick, innerStateShow]);

  const Trigger = customDropdownTrigger || DefaultTrigger;
  const isCustomTriggerMode = !!customDropdownTrigger;

  const handleCloseOnItemClick = () => {
    if (closeOnItemClick) {
      closeDropdownHandler();
    }
  };

  const renderChildrenItems = () => {
    return React.Children.toArray(children)
      .filter(Boolean)
      .map((child, index) => {
        return React.cloneElement(child, {
          handleClose: handleCloseOnItemClick,
          itemindex: index
        });
      });
  };

  const transitions = useTransition(dropdownOpenned, null, {
    from: { position: "absolute", opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0 },
    duration: 200
  });

  return (
    <div
      ref={dropdownRef}
      className={classNames(styles.dropdownWrapper, className)}
      {...props}>
      <Trigger
        title={title}
        onClick={handleTriggerClick}
        open={dropdownOpenned}
      />
      {isCustomTriggerMode && (
        <div
          className={classNames(
            styles.arrowAnchor,
            dropdownOpenned && styles.open,
            dropdownArrowClassName
          )}
        />
      )}
      {dropdownOpenned &&
        transitions.map(
          ({ item, key, props }) =>
            item && (
              <animated.ul
                className={classNames(styles.dropdownList, itemsListClassName)}
                key={key}
                style={props}>
                {renderChildrenItems()}
              </animated.ul>
            )
        )}
    </div>
  );
};

DefaultTrigger.propTypes = {
  onClick: PropTypes.func.isRequired,
  title: PropTypes.string,
  open: PropTypes.bool
};

Dropdown.propTypes = {
  className: PropTypes.string,
  itemsListClassName: PropTypes.string,
  dropdownArrowClassName: PropTypes.string,
  customDropdownTrigger: PropTypes.node,
  title: PropTypes.string,
  show: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onDropdownClick: PropTypes.func,
  closeOnOutsideClick: PropTypes.bool,
  closeOnItemClick: PropTypes.bool
};

Dropdown.defaultProps = {
  closeOnOutsideClick: true,
  closeOnItemClick: true
};

export default Dropdown;
