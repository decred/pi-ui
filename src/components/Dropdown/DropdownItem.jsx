import PropTypes from "prop-types";
import React from "react";
import styles from "./styles.module.css";
import { classNames } from "../../utils";

const DropdownItem = ({
  handleClose,
  onClick,
  className,
  children,
  ...props
}) => {
  const handleClick = () => {
    handleClose && handleClose();
    onClick && onClick();
  };
  return (
    <li
      onClick={handleClick}
      className={classNames(styles.dropdownItem, className)}
      {...props}>
      {children}
    </li>
  );
};

DropdownItem.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,
  handleClose: PropTypes.func,
  onClick: PropTypes.func,
  style: PropTypes.object,
};

export default DropdownItem;
