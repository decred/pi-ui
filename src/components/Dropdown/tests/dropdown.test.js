import React from "react";
import Dropdown from "../Dropdown";
import DropdownItem from "../DropdownItem";
import { create } from "react-test-renderer";
import { fireEvent, render, wait } from "@testing-library/react";

describe("Dropdown component", () => {
  test("Matches snapshot", () => {
    const dropdown = create(
      <Dropdown title="test">
        <DropdownItem>test1</DropdownItem>
        <DropdownItem>test2</DropdownItem>
      </Dropdown>
    );
    expect(dropdown.toJSON()).toMatchSnapshot();
  });
  test("Items list show on click", async () => {
    const itemMockClick = jest.fn();
    const { getByTestId, queryByTestId, getByText, queryByText } = render(
      <Dropdown title="test">
        <DropdownItem onClick={itemMockClick}>test1</DropdownItem>
        <DropdownItem>test2</DropdownItem>
      </Dropdown>
    );
    const trigger = queryByTestId("trigger");
    const itemsList = queryByTestId("items-list");
    expect(trigger).toBeTruthy();
    expect(itemsList).toBeFalsy();
    fireEvent.click(getByTestId("trigger"));
    const updatedItemsList = queryByTestId("items-list");
    expect(updatedItemsList).toBeTruthy();
    const dropdownItem = queryByText(/test1/i);
    expect(dropdownItem).toBeTruthy();
    fireEvent.click(getByText("test1"));
    expect(itemMockClick).toBeCalled();
    const afterClickItemsList = queryByTestId("items-list");
    expect(afterClickItemsList).toBeFalsy();
  });
});
