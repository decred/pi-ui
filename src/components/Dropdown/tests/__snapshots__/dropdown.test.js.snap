// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Matches snapshot 1`] = `
<div
  className="dropdownWrapper"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="headerWrapper"
    data-testid="trigger"
    onClick={[Function]}
  >
    <span
      className="dropdownHeader"
    >
      test
    </span>
    <div
      className="arrowAnchor"
      onClick={[Function]}
    />
  </div>
</div>
`;
